source/agora/common/ManagedDatabase.d

```diff
@@ -0,0 +1,106 @@
+/*******************************************************************************
+
+    Contains a managed wrapper around SQLite databases.
+
+    This class ensures that all D2SQLite databases handled by the thread
+    are cleanly destroyed after the thread shuts down.
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.ManagedDatabase;
+
+import d2sqlite3.database;
+import d2sqlite3.results;
+import d2sqlite3.sqlite3;
+
+/// Ditto
+public class ManagedDatabase
+{
+    /// ManagedDatabase managed by the current thread
+    private static Database*[] thread_dbs;
+
+    /// Close all database handles
+    static ~this ()
+    {
+        foreach (db; thread_dbs)
+            db.close();
+    }
+
+    /// Pointer to the database handle
+    private Database* database;
+
+    /// Subtype
+    public alias getDatabase this;
+
+    /***************************************************************************
+
+        Constructs a managed database
+
+        Params:
+            path = path to the db on disk, or :memory: for an in-memory database
+            flags = SQLite read / write flags
+
+    ***************************************************************************/
+
+    public this (string path,
+        int flags = SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE)
+    {
+        this.database = new Database(path, flags);
+        thread_dbs ~= this.database;
+    }
+
+    /***************************************************************************
+
+        @safe wrapper around db.execute()
+
+        The ResultRange @system destructor will be called within the context of
+        this @trusted function, enabling the calling code to be @safe.
+
+        Params:
+            Args = inferred
+            T = inferred from the optionally provided delegate, its return type
+            sql = the sql statement to execute
+            dg = if provided, a delegate which will be called on the ResultRange
+                 of the SQLite execute() result. If the delegate has a return
+                 value, it will be returned in the execute() call.
+            args = any arguments to bind to
+
+    ***************************************************************************/
+
+    public void execute (Args...)(string sql, auto ref Args args) @trusted
```

Comment:
Yeah that's a `@safe` violation alright.

Reply:
Yeah I realized this as soon as I made the PR.

Reply:
s/ made the PR / added the commit